src/runner.rs

```diff
@@ -6,511 +6,183 @@
-    pub error_count: u32,
-}
-
-impl<'a> Runner<'a> {
-
-    #[cfg_attr(feature="clippy", allow(redundant_closure))]
-    fn run_test(mut test_function: Box<TestFunction>)
-                -> ExampleResult {
-
-        use std::panic::{catch_unwind, AssertUnwindSafe};
-        use example_result;
-
-        let res = catch_unwind(AssertUnwindSafe(|| test_function()));
-
-        res.unwrap_or_else(|_| example_result::FAILED_RES)
+impl<'a, T> Runner<'a, T>
+    where T: 'a + Clone + ::std::fmt::Debug
+{
+    pub fn run(mut self) -> ContextReport {
+        let suite = mem::replace(&mut self.suite, None).expect("Expected context");
+        panic::set_hook(Box::new(|_panic_info| {
+            // silently swallows panics
+        }));
+        self.visit(&suite);
+        let _ = panic::take_hook();
+        self.report
     }
 
-    fn run_and_recurse(report: &mut TestReport,
-                       child_ctx: &mut Context,
-                       handlers: &mut Handlers)
-                       -> ExampleResult {
-
-        use std::mem;
-        let mut result = example_result::SUCCESS_RES;
-
-        // As Runner::run consomes the runner, we can deconstruct the context tree as we travel
-        // among it.
-        // This effectively takes ownership of the context node and replace it with an empty one.
-        let ctx = mem::replace(child_ctx, Context::default());
-
-        let tests = ctx.tests;
-        let mut before_functions = ctx.before_each;
-        let mut after_functions = ctx.after_each;
-
-
-        for mut test_function in tests {
-            let test_res = {
-                for before_function in before_functions.iter_mut() {
-                    before_function()
-                }
-                let res = match test_function {
-                    Testable::Test(name, test_function) => {
-                        handlers.broadcast(&Event::StartTest(name));
-                        let res = Runner::run_test(test_function);
-                        handlers.broadcast(&Event::EndTest(res));
-                        res
-                    }
-                    Testable::Describe(ref name, ref mut desc) => {
-                        handlers.broadcast(&Event::StartDescribe(name.clone()));
-                        let res = Runner::run_and_recurse(report, desc, handlers);
-                        handlers.broadcast(&Event::EndDescribe);
-                        res
-                    }
-                };
-                for after_function in after_functions.iter_mut() {
-                    after_function()
-                }
-                res
-            };
-
-            report.total_tests += 1;
-
-            if test_res.is_ok() {
-                report.success_count += 1;
-            } else {
-                report.error_count += 1
-            }
-
-            result = test_res.or(result);
+    pub fn run_or_exit(self) {
+        if self.run().failed > 0 {
+            ::std::process::exit(101);
```

Comment:
Excellent ! I suggest you put this link in comment just before the `process::exit` line. Something like that:
```rust
   pub fn run_or_exit(self) {
        if self.run().failed > 0 {
             // XXX Cargo test failure returns 101. Ref: https://www.reddit.com/r/rust/comments/3qdvez/guarantees_about_exit_codes/cwf46cb/

            ::std::process::exit(101);
```